### DISCUSSION TITLE
Open app on specific workspace EXCEPT

### DISCUSSION THREAD

**Discussion by mobsenpai** [hyprwm/Hyprland#5521] [Q&A]
How can I say open firefox on workspace 1 except ~GtkFileChooser~ Upload file/folder dialog window.

> Currently I've set this rule and this makes the file chooser to also open on workspace 1 so, what can I do to stop that? Is there any "except" keyword?
```
workspace 1 silent, class:^(firefox)$
```

**Comment by 0xFMD** [comment]
Check the title with ``` hyprctl clients ```
like
``` windowrulev2 =  workspace 9 silent, class:^(firefox)$,title:^(GtkFileChooser)$```

**Reply by mobsenpai** [reply]
Hmm but If I don't wanna assign any workspace to gtkfilechooser then? Is there a rule to clear any workspace rules or anything?

**Reply by 0xFMD** [reply]
I didn't get it. Maybe, you can open "**GtkFileChooser**" on special workspace or use "**[_Relative workspace_](https://wiki.hyprland.org/Configuring/Dispatchers/#workspaces)**" 

```
windowrulev2 = workspace 1 silent, class:^(firefox)$
windowrulev2 = workspace r+1 silent, class:^(firefox)$,title:^(GtkFileChooser)$

```

**Comment by mobsenpai** [comment]
Hello @MysticBytes786, I apologize for any confusion earlier. I want to clarify my request regarding the Firefox window behavior on Hyprland. I have set up a rule to open Firefox on Workspace 1. However, I've noticed an issue where if I move the Firefox window to another workspace, say Workspace 2, and try to upload a file on a website, the file chooser window, which has a class of "GtkFileChooser," inherits the Workspace 1 rule and opens on Workspace 1 instead of the current workspace (Workspace 2). Ideally, I would like the file chooser window to be independent of the workspace rule and open on the same workspace as the Firefox window. This way, the file chooser window will appear on Workspace 2 when I move the Firefox window there. Here's an image of the file chooser window I'm referring to: image I hope this clarifies my request. I apologize for any confusion caused by my previous message, and I appreciate your help in resolving this issue. Thank you for your patience, and I apologize again if I come across as rude due to my non-native English proficiency.

> firefox window is on workspace 2 and the file chooser opened on workspace 1
![image](https://github.com/hyprwm/Hyprland/assets/92603465/35836282-30d0-4298-9b8f-a295e654cf30)

**Reply by 0xFMD** [reply]
Hmm, maybe this will work.

```
windowrulev2 = workspace 1 silent, class:^(firefox)$
windowrulev2 = workspace r, class:^(firefox)$, title:^(GtkFileChooser)$
```

**Reply by mobsenpai** [reply]
"The workspace r command is indeed functional. However, I would like to correct a detail in my previous request. I apologize for not verifying earlier that the GtkFileChooser class is accurate in Xorg. In Wayland, there is no separate class for this; instead, the title of the Firefox window changes dynamically based on the websites being accessed. For instance, the title may be "Select Folder to Upload" or "Upload Folder," depending on the website. I have confirmed this by testing with hyprctl clients. Given that the window title changes dynamically, what would be the appropriate approach to tackle this situation while using the workspace r command? I believe that I need to specify the correct window title. Is that correct?"

**Reply by 0xFMD** [reply]
You can use [RegEx](https://wiki.hyprland.org/Configuring/Window-Rules/#window-rules-v2)

**Comment by mobsenpai** [comment]
Thank you for your prompt and helpful responses.
 I have been able to use the following command to achieve the desired result:
~`"workspace 1 silent, class:^(firefox)$, !title:^(.*Upload*.)$"`~

```
workspace 1 silent, class:^(firefox)$
workspace r silent, class:^(firefox)$, title:^$(.*Upload*.)$
```

**Reply by mobsenpai** [reply]
@MysticBytes786 I thought I had it when I used !title but later did I know that there is no syntax including ! right? that's what I originally meant when I said except I thought there was something like !title or something, but there wasn't, so I changed it to current answer and it's working fine but I don't think this method is very reliable cuz it's just a title there may be other websites that have upload in title. So I don't know, for now let's keep this.
